File: app/src/main/java/cenergy/central/com/pwb_store/activity/CompareActivity.java
package cenergy.central.com.pwb_store.activity;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import cenergy.central.com.pwb_store.R;
import cenergy.central.com.pwb_store.adapter.interfaces.CompareItemListener;
import cenergy.central.com.pwb_store.fragment.CompareFragment;
import cenergy.central.com.pwb_store.manager.ApiResponseCallback;
import cenergy.central.com.pwb_store.manager.HttpManagerMagento;
import cenergy.central.com.pwb_store.manager.bus.event.CompareDetailBus;
import cenergy.central.com.pwb_store.manager.preferences.PreferenceManager;
import cenergy.central.com.pwb_store.model.APIError;
import cenergy.central.com.pwb_store.model.CacheCartItem;
import cenergy.central.com.pwb_store.model.CartItem;
import cenergy.central.com.pwb_store.model.CompareProduct;
import cenergy.central.com.pwb_store.model.body.CartBody;
import cenergy.central.com.pwb_store.model.body.CartItemBody;
import cenergy.central.com.pwb_store.realm.DatabaseListener;
import cenergy.central.com.pwb_store.realm.RealmController;
import cenergy.central.com.pwb_store.utils.DialogUtils;
import cenergy.central.com.pwb_store.view.PowerBuyShoppingCartView;

/**
 * Created by napabhat on 7/26/2017 AD.
 */

public class CompareActivity extends AppCompatActivity implements CompareItemListener, PowerBuyShoppingCartView.OnClickListener {

    Toolbar mToolbar;
    PowerBuyShoppingCartView mBuyShoppingCartView;
    private PreferenceManager preferenceManager;
    private ProgressDialog mProgressDialog;
    private RealmController database = RealmController.getInstance();

    @Subscribe
    public void onEvent(CompareDetailBus compareDetailBus) {
//        Intent intent = new Intent(this, ProductDetailActivity.class);
//        intent.putExtra(ProductDetailActivity.ARG_PRODUCT_ID, compareDetailBus.getProductCompareList().getProductId());
//        ActivityCompat.startActivity(this, intent,
//                ActivityOptionsCompat
//                        .makeScaleUpAnimation(compareDetailBus.getView(), 0, 0, compareDetailBus.getView().getWidth(), compareDetailBus.getView().getHeight())
//                        .toBundle());

        Intent intent = new Intent(this, ProductDetailActivity.class);
        intent.putExtra(ProductDetailActivity.ARG_PRODUCT_SKU, compareDetailBus.getCompareProduct().getSku());
        ActivityCompat.startActivity(this, intent,
                ActivityOptionsCompat
                        .makeScaleUpAnimation(compareDetailBus.getView(), 0, 0, compareDetailBus.getView().getWidth(), compareDetailBus.getView().getHeight())
                        .toBundle());
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_compare);

        preferenceManager = new PreferenceManager(this); // get pref

        initView();

        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction
                .replace(R.id.container, CompareFragment.newInstance())
                .commit();
    }

    private void initView() {
        mToolbar = findViewById(R.id.toolbar);
        mBuyShoppingCartView = findViewById(R.id.shopping_cart_compare);
        setSupportActionBar(mToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
        }
        mToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
        mBuyShoppingCartView.setListener(this);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        EventBus.getDefault().register(this);
        updateShoppingCartBadge();
    }

    @Override
    protected void onPause() {
        EventBus.getDefault().unregister(this);
        super.onPause();
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_CANCELED) {
            finish();
        }
    }

    @Override
    public void onShoppingCartClick(View view) {
        if (database.getCacheCartItems().size() > 0) {
            ShoppingCartActivity.Companion.startActivity(this, view, preferenceManager.getCartId());
        } else {
            showAlertDialog("", getResources().getString(R.string.not_have_products_in_cart));
        }
    }

    private void showAlertDialog(String title, String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.AlertDialogTheme)
                .setMessage(message)
                .setPositiveButton(getString(R.string.ok_alert), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        if (mProgressDialog != null && mProgressDialog.isShowing()) {
                            mProgressDialog.dismiss();
                        }
                    }
                });

        if (!TextUtils.isEmpty(title)) {
            builder.setTitle(title);
        }
        builder.show();
    }

    private void showProgressDialog() {
        if (mProgressDialog == null) {
            mProgressDialog = DialogUtils.createProgressDialog(this);
            mProgressDialog.show();
        } else {
            if (!mProgressDialog.isShowing()) {
                mProgressDialog.show();
            }
        }
    }

    @Override
    public void onPointerCaptureChanged(boolean hasCapture) {

    }

    // region {@link {Implement CompareItemListener}
    @Override
    public void onClickShoppingCart(CompareProduct compareProduct) {
        String cartId = preferenceManager.getCartId();
        if (preferenceManager.getCartId() != null) {
            Log.d("ProductDetail", "has cart id");
            addProductToCart(cartId, compareProduct);
        } else {
            Log.d("ProductDetail", "new cart id");
            retrieveCart(compareProduct);
        }
    }
    // endregion

    private void retrieveCart(final CompareProduct compareProduct) {
        showProgressDialog();
        HttpManagerMagento.Companion.getInstance(this).getCart(new ApiResponseCallback<String>() {
            @Override
            public void success(@Nullable String cartId) {
                if (cartId != null) {
                    preferenceManager.setCartId(cartId);
                    addProductToCart(cartId, compareProduct);
                }
            }

            @Override
            public void failure(@NotNull APIError error) {
                if (mProgressDialog != null && mProgressDialog.isShowing()) {
                    mProgressDialog.dismiss();
                }
                showAlertDialog("", error.getErrorUserMessage());
            }
        });
    }

    private void addProductToCart(final String cartId, final CompareProduct compareProduct) {
        showProgressDialog();
        CartItemBody cartItemBody = new CartItemBody(new CartBody(cartId, compareProduct.getSku(), 1)); // default add qty 1
        HttpManagerMagento.Companion.getInstance(this).addProductToCart(cartId, cartItemBody, new ApiResponseCallback<CartItem>() {
            @Override
            public void success(@Nullable CartItem cartItem) {
                saveCartItem(cartItem, compareProduct);
                if (mProgressDialog != null && mProgressDialog.isShowing()) {
                    mProgressDialog.dismiss();
                }
            }

            @Override
            public void failure(@NotNull APIError error) {
                if (mProgressDialog != null && mProgressDialog.isShowing()) {
                    mProgressDialog.dismiss();
                }

                if (error.getErrorCode().equals(String.valueOf(APIError.INTERNAL_SERVER_ERROR))) {
                    showClearCartDialog();
                } else {
                    showAlertDialog("", error.getErrorMessage());
                }
            }
        });
    }

    private void saveCartItem(CartItem cartItem, final CompareProduct compareProduct) {
        database.saveCartItem(CacheCartItem.asCartItem(cartItem, compareProduct), new DatabaseListener() {
            @Override
            public void onSuccessfully() {
                updateShoppingCartBadge();
                Toast.makeText(CompareActivity.this, getString(R.string.added_to_cart), Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onFailure(Throwable error) {
                mProgressDialog.dismiss();
                if (error != null) {
                    showAlertDialog("", error.getMessage());
                }
            }
        });
    }

    private void updateShoppingCartBadge() {
        int count = 0;
        List<CacheCartItem> items = database.getCacheCartItems();
        for (CacheCartItem item : items) {
            if (item.getQty() != null) {
                count += item.getQty();
            }
        }
        mBuyShoppingCartView.setBadgeCart(count);
        Log.d("ProductDetail", "count shopping badge" + count);
    }


    private void showClearCartDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.AlertDialogTheme)
                .setMessage(getString(R.string.title_clear_cart))
                .setPositiveButton(getString(R.string.ok_alert), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        clearCart(); // clear item cart
                        updateShoppingCartBadge(); // update ui
                    }
                });
        builder.show();
    }

    private void clearCart() {
        database.deleteAllCacheCartItem();
        preferenceManager.clearCartId();
    }

}

File: app/src/main/java/cenergy/central/com/pwb_store/activity/CompareActivity.kt
package cenergy.central.com.pwb_store.activity

import android.content.Intent
import android.net.NetworkInfo
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.ActivityOptionsCompat
import android.support.v7.app.AlertDialog
import android.support.v7.widget.Toolbar
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import cenergy.central.com.pwb_store.R
import cenergy.central.com.pwb_store.activity.interfaces.CompareProtocol
import cenergy.central.com.pwb_store.adapter.interfaces.CompareItemListener
import cenergy.central.com.pwb_store.extensions.isProductInStock
import cenergy.central.com.pwb_store.fragment.CompareFragment
import cenergy.central.com.pwb_store.helpers.DialogHelper
import cenergy.central.com.pwb_store.manager.ApiResponseCallback
import cenergy.central.com.pwb_store.manager.HttpManagerMagento
import cenergy.central.com.pwb_store.manager.api.CompareAPI
import cenergy.central.com.pwb_store.manager.bus.event.CompareDeleteBus
import cenergy.central.com.pwb_store.manager.bus.event.CompareDetailBus
import cenergy.central.com.pwb_store.manager.preferences.AppLanguage
import cenergy.central.com.pwb_store.model.APIError
import cenergy.central.com.pwb_store.model.CacheCartItem
import cenergy.central.com.pwb_store.model.CartItem
import cenergy.central.com.pwb_store.model.CompareProduct
import cenergy.central.com.pwb_store.model.body.CartItemBody
import cenergy.central.com.pwb_store.model.response.CompareProductResponse
import cenergy.central.com.pwb_store.realm.DatabaseListener
import cenergy.central.com.pwb_store.realm.RealmController
import cenergy.central.com.pwb_store.view.LanguageButton
import cenergy.central.com.pwb_store.view.NetworkStateView
import cenergy.central.com.pwb_store.view.PowerBuyShoppingCartView
import kotlinx.android.synthetic.main.activity_compare.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

class CompareActivity : BaseActivity(), CompareItemListener, PowerBuyShoppingCartView.OnClickListener, CompareProtocol {

    private lateinit var mToolbar: Toolbar
    private lateinit var mBuyShoppingCartView: PowerBuyShoppingCartView
    private val database = RealmController.getInstance()
    private lateinit var languageButton: LanguageButton
    private lateinit var networkStateView: NetworkStateView
    private var compareProducts: List<CompareProduct> = arrayListOf()
    private var compareProductDetailList: List<CompareProductResponse> = arrayListOf()

    @Subscribe
    fun onEvent(compareDetailBus: CompareDetailBus) {
        val intent = Intent(this, ProductDetailActivity::class.java)
        intent.putExtra(ProductDetailActivity.ARG_PRODUCT_SKU, compareDetailBus.compareProduct.sku)
        ActivityCompat.startActivity(this, intent,
                ActivityOptionsCompat
                        .makeScaleUpAnimation(compareDetailBus.view, 0, 0, compareDetailBus.view.width, compareDetailBus.view.height)
                        .toBundle())
    }

    @Subscribe
    fun onEvent(compareDeleteBus: CompareDeleteBus) {
        database.deleteCompareProduct(compareDeleteBus.compareProduct.sku)
        retrieveCompareProduct()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_compare)
        languageButton = findViewById(R.id.switch_language_button)
        networkStateView = findViewById(R.id.networkStateView)
        handleChangeLanguage()
        initView()
        retrieveCompareProduct()
    }

    override fun getCompateProducts(): List<CompareProduct> = compareProducts
    override fun getCompareProductDetailList() = compareProductDetailList

    private fun retrieveCompareProduct() {
        showProgressDialog()
        CompareAPI().retrieveCompareProduct(this, getSKUs(), object : ApiResponseCallback<List<CompareProductResponse>> {
            override fun success(response: List<CompareProductResponse>?) {
                compareProductDetailList = response ?: arrayListOf()
                startCompareFragment()
                dismissProgressDialog()
            }

            override fun failure(error: APIError) {
                startCompareFragment()
                dismissProgressDialog()
            }
        })
    }

    private fun getSKUs(): String {
        // update compare product list
        compareProducts = database.compareProducts

        val productSKUs = arrayListOf<String>()
        compareProducts.forEach {
            productSKUs.add(it.sku)
        }
        return productSKUs.joinToString(",")
    }

    private fun startCompareFragment() {
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction
                .replace(R.id.container, CompareFragment.newInstance(), CompareFragment.tag)
                .commit()
    }

    private fun initView() {
        mToolbar = findViewById(R.id.toolbar)
        mBuyShoppingCartView = findViewById(R.id.shopping_cart_compare)
        setSupportActionBar(mToolbar)
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayShowTitleEnabled(false)
        }
        mToolbar.setNavigationOnClickListener { finish() }
        mBuyShoppingCartView.setListener(this)
    }

    override fun onResume() {
        super.onResume()
        EventBus.getDefault().register(this)
        updateShoppingCartBadge()
    }

    override fun onPause() {
        EventBus.getDefault().unregister(this)
        super.onPause()
    }

    // region language
    override fun onChangedLanguage(lang: AppLanguage) {
        super.onChangedLanguage(lang)
        retrieveCompareProduct()
    }

    override fun getSwitchButton(): LanguageButton? = languageButton
    // endregion

    // region network
    override fun onNetworkStateChange(state: NetworkInfo.State) {
        super.onNetworkStateChange(state)
        // isConnected?
        if (currentState == NetworkInfo.State.CONNECTED && forceRefresh) {
            retrieveCompareProduct()
        }
    }

    override fun getStateView(): NetworkStateView? = networkStateView
    // endregion

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == REQUEST_UPDATE_LANGUAGE) {
            // check language
            if (getSwitchButton() != null) {
                getSwitchButton()!!.setDefaultLanguage(preferenceManager.getDefaultLanguage())
            }

            if (resultCode == ShoppingCartActivity.RESULT_UPDATE_PRODUCT) {
                retrieveCompareProduct()
            }
        }
    }

    override fun onShoppingCartClick(view: View) {
        if (database.cacheCartItems!!.size > 0) {
            ShoppingCartActivity.startActivity(this, preferenceManager.cartId)
        } else {
            showAlertDialog("", resources.getString(R.string.not_have_products_in_cart))
        }
    }

    private fun showAlertDialog(title: String, message: String) {
        val builder = AlertDialog.Builder(this, R.style.AlertDialogTheme)
                .setMessage(message)
                .setPositiveButton(getString(R.string.ok_alert)) { _, _ ->
                    dismissProgressDialog()
                }

        if (!TextUtils.isEmpty(title)) {
            builder.setTitle(title)
        }
        builder.show()
    }

    private fun showOKAlertDialog(message: String) {
        val builder = AlertDialog.Builder(this, R.style.AlertDialogTheme)
                .setMessage(message)
                .setPositiveButton(getString(R.string.ok_alert)) { _, _ ->
                    clearAllCompareProducts()
                }
                .setNegativeButton(getString(R.string.no)) { dialog, _ ->
                    dialog.dismiss()
                }
        builder.show()
    }

    private fun showProgressDialog() {
        loadingProgressBar?.visibility = View.VISIBLE
    }

    // region {@link {Implement CompareItemListener}
    override fun onClickAddToCart(compareProduct: CompareProduct) {
        showProgressDialog()
        val cartId = preferenceManager.cartId
        if (cartId != null) {
            addProductToCart(cartId, compareProduct)
        } else {
            retrieveCart(compareProduct)
        }
    }

    override fun onClearAllProductCompare() {
        if (getSKUs().isNotEmpty()){
            showOKAlertDialog(getString(R.string.text_comfirm_clear_all))
        }
    }
    // endregion

    private fun clearAllCompareProducts(){
        database.deleteAllCompareProduct()
        val fragment = supportFragmentManager.findFragmentByTag(CompareFragment.tag)
        if (fragment != null) {
            fragment as CompareFragment
            fragment.compareProducts = arrayListOf()
            fragment.compareProductDetailList = arrayListOf()
            fragment.updateCompareList()
        }
    }

    private fun retrieveCart(compareProduct: CompareProduct) {
        showProgressDialog()
        HttpManagerMagento.getInstance(this).getCart(object : ApiResponseCallback<String?> {
            override fun success(response: String?) {
                if (response != null) {
                    preferenceManager.setCartId(response)
                    addProductToCart(response, compareProduct)
                }
            }

            override fun failure(error: APIError) {
                showAlertDialog("", error.errorUserMessage)
            }
        })
    }

    private fun addProductToCart(cartId: String, compareProduct: CompareProduct) {
        val cartItemBody = CartItemBody.create(cartId, compareProduct.sku, arrayListOf())
        HttpManagerMagento.getInstance(this).addProductToCart(cartId, cartItemBody, object : ApiResponseCallback<CartItem> {
            override fun success(response: CartItem?) {
                runOnUiThread {
                    saveCartItem(response, compareProduct)
                }
            }

            override fun failure(error: APIError) {
                dismissProgressDialog()
                if (error.errorCode == APIError.INTERNAL_SERVER_ERROR.toString()) {
                    showClearCartDialog()
                } else {
                    DialogHelper(this@CompareActivity).showErrorDialog(error)
                }
            }
        })
    }

    private fun saveCartItem(cartItem: CartItem?, compareProduct: CompareProduct) {
        cartItem ?: return
        database.saveCartItem(CacheCartItem.asCartItem(cartItem, compareProduct), object : DatabaseListener {
            override fun onSuccessfully() {
                dismissProgressDialog()
                updateShoppingCartBadge()
                Toast.makeText(this@CompareActivity, getString(R.string.added_to_cart),
                        Toast.LENGTH_SHORT).show()

                updateCompareList(compareProduct)
            }

            override fun onFailure(error: Throwable) {
                showAlertDialog("", error.message ?: getString(R.string.some_thing_wrong))
            }
        })
    }

    private fun updateCompareList(compareProduct: CompareProduct) {
        // is no have in stock? force update
        if (!this@CompareActivity.isProductInStock(compareProduct)) {
            val fragment = supportFragmentManager.findFragmentByTag(CompareFragment.tag)
            if (fragment != null) {
                (fragment as CompareFragment).updateCompareList()
            }
        }
    }

    private fun updateShoppingCartBadge() {
        var count = 0
        val items = database.cacheCartItems
        for (item in items!!) {
            if (item.qty != null) {
                count += item.qty!!
            }
        }
        mBuyShoppingCartView.setBadgeCart(count)
    }

    private fun clearCart() {
        database.deleteAllCacheCartItem()
        preferenceManager.clearCartId()
    }

    private fun showClearCartDialog() {
        val builder = AlertDialog.Builder(this, R.style.AlertDialogTheme)
                .setCancelable(false)
                .setMessage(getString(R.string.title_clear_cart))
                .setPositiveButton(getString(R.string.ok_alert)) { dialog, _ ->
                    clearCart() // clear item cart
                    updateShoppingCartBadge() // update ui
                    dialog.dismiss()
                }
        builder.show()
    }

    private fun dismissProgressDialog() {
        loadingProgressBar?.visibility = View.INVISIBLE
    }
}
